### PR TITLE
option additions

### PR DESCRIPTION
**Pull request by lpil**
I've added some new functions to option. This is unfinished, see the comments.

I've found rebar3 to crash a lot while attempting to run eunit, I was unable to determine a pattern as to what caused it. Sometimes deleting _build and recompiling would fix the problem. Is there any own problems with rebar3?

### CODE REVIEW

**Comment by j14159** [comment]
> I've  found rebar3 to crash a lot while attempting to run eunit, I was unable  to determine a pattern as to what caused it. Sometimes deleting _build  and recompiling would fix the problem. Is there any own problems with  rebar3?

I ran into this as well, it's an issue in the exhaustiveness checker.  When one module depends on a type from another which has already been compiled, the exhaustiveness checker can't find the module from which to retrieve the type.

E.g.

    module A
    export type a
    type a = A int

and

    module B
    import type A.a
    let addXToA x A y = A (x + y)

if `A` is compiled from a previous run and then we add `B` (or make changes to it), `A` won't be in the list of modules for the exhaustiveness checker and a list lookup will fail.  I'll file this issue under Alpaca itself now so we don't lose it :)

I suspect we need to add an annotation to the beam files with erlang terms for all the exported ADTs unless it's already there and I'm overlooking it.

Will copy/paste this into an Alpaca issue.

File: src/option.alp [hunk 1]
```diff
@@ -60,4 +62,96 @@ test "mapping double function should double the option" =
...
+
+test "flatten of Some None" =
+  assert.equal (flatten (Some None)) None
+
+test "flatten of None" =
+  assert.equal (flatten None) None
+
+-- TODO
+-- I feel we should be able to defined this like so. But it does not type check.
+-- Is this a bug in the compiler?
```
Comment (j14159) [inline src/option.alp]: Looks like a bug to me.  Would you mind filing this against Alpaca?  Thanks!

File: src/option.alp [hunk 2]
```diff
@@ -60,4 +62,96 @@ test "mapping double function should double the option" =
...
+    ```
+    flatten (Some None) == None
+    ```
+    ```
+    flatten None == None
+    ```
+-}
+
+-- TODO
+-- Is this typespec correct? It does not type-check if uncommented.
```
Comment (j14159) [inline src/option.alp]: Per below I think this is a bug in Alpaca.

Comment (j14159) [reply src/option.alp]: (thanks for finding it!)

**Comment by lpil** [comment]
Updated the PR

**Comment by j14159** [comment]
Thanks!  Looks good to me though if you don't mind I'm going to leave this open for a day or two for:
- @tsloughter and/or @lepoetemaudit to have a chance to take a look (iirc they both have slightly stronger opinions about standard lib stuff than I and have thought about it more)
- me to add a Travis CI build.  I think it might be good if we built `alpaca_lib` with the current Alpaca release and current master branch of Alpaca, and on both OTP 19.3 and 20.0.  Thoughts on this?  Might be nice to have 4 Travis CI badges to show the state of things (this is a vague/rough idea, don't know if it's possible).

**Comment by lpil** [comment]
CI sounds good to me :)

I have a mild preference for Circle CI 2 over Travis as I find the Docker based system easier to work with than the Travis system but CI in general is a win by me.

**Comment by lpil** [comment]
When can we merge this? I'd like to make use of this module in my next contributions.

**Comment by lepoetemaudit** [comment]
I have no objections, all looks good to me.

**Comment by j14159** [comment]
Merging per @lepoetemaudit 's agreement, will add CI today.

**Comment by lpil** [comment]
Thanks all! 💜

**Pull request merged**